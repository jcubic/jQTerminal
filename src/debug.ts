const warnings: Array<string> = [];

export function warn(msg: string) {
    msg = '[Pure Terminal] ' + msg;
    if (warnings.indexOf(msg) === -1) {
        warnings.push(msg);
        /* eslint-disable */
        if (console) {
            if (console.warn) {
                console.warn(msg);
            } else if (console.log) {
                console.log(msg);
            }
            /* eslint-enable */
        } else {
            // prevent catching in outer try..catch
            setTimeout(function() {
                throw new Error('WARN: ' + msg);
            }, 0);
        }
    }
}
