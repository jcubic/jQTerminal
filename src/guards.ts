export function is_string(arg: unknown): arg is string {
    return typeof arg === 'string';
}

export function is_function(arg: unknown): arg is () => unknown {
    return typeof arg === 'function';
}

export function is_array(arg: unknown): arg is Array<unknown> {
    return Array.isArray(arg);
}

export function is_integer(arg: unknown): arg is number {
    return Number.isInteger(arg);
}

export type Hash_Command = [number, number, string];

export function is_hash_command(arg: unknown): arg is Hash_Command {
    if (!is_array(arg)) {
        return false;
    }
    if (arg.length !== 3) {
        return false;
    }
    if (!(is_integer(arg[0]) && is_integer(arg[1]))) {
        return false;
    }
    return is_string(arg[2]);

}
