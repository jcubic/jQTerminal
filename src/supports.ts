import { mobile_re, tablet_re } from './const';

const agent = (self.navigator || window.navigator).userAgent;
const is_IE = /MSIE|Trident/.test(agent) || /rv:11.0/i.test(agent);
const is_IEMobile = /IEMobile/.test(agent);

export const is_ch_unit_supported = (function() {
    if (is_IE && !is_IEMobile) {
        return false;
    }
    if (typeof document === 'undefined') {
        return true; // run without browser context
    }
    const div = document.createElement('div');
    div.style.width = '1ch';
    return div.style.width === '1ch';
})();

export const is_css_variables_supported = self.CSS && self.CSS.supports &&
    self.CSS.supports('--fake-var', '0');

export const is_android = navigator.userAgent.toLowerCase().indexOf('android') !== -1;


export const is_key_native = (function is_key_native() {
    if (!('KeyboardEvent' in self && 'key' in self.KeyboardEvent.prototype)) {
        return false;
    }
    const proto = self.KeyboardEvent.prototype;
    const get = Object.getOwnPropertyDescriptor(proto, 'key')?.get;
    return !!get?.toString().match(/\[native code\]/);
})();

export const is_browser = globalThis.window === globalThis;

export const is_mobile = (function(a) {
    let check = false;
    if (mobile_re.test(a) || tablet_re.test(a.substr(0, 4))) {
        check = true;
    }
    // detect iPad 13
    // ref: https://stackoverflow.com/a/57924983/387194s
    if (navigator.platform === 'MacIntel' && navigator.maxTouchPoints > 1) {
        return true;
    }
    return check;
})(navigator.userAgent || navigator.vendor || (self as any).opera);
