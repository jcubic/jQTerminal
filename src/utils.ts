export const escape_regex = (str: string) => {
    const special = /([-\\^$[\]()+{}?*.|])/g;
    return str.replace(special, '\\$1');
};
