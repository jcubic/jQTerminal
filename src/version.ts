const VERSION = 'v. 3.0.0';

export default VERSION;
