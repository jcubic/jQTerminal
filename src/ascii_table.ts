import wcwidth from 'wcwidth';
import { strip } from './formatters/utils';

type Field = string | number | boolean | RegExp;
type Table = Array<Array<Field>>;
type StringTable = Array<Array<string>>;

export function strlen(str: string) {
    return wcwidth(strip(str));
}

function format_argument(arg: Field): string {
    return expand_tabs(arg.toString().replace(/\r/g, ''));
}

function adjust_table(array: Table): StringTable {
    array = [...array];
    // adjustment for newlines based on cviejo
    // ref: https://stackoverflow.com/a/35115703/387194
    for (let i = array.length - 1; i >= 0; i--) {
        const row = array[i];
        const stacks: StringTable = [];
        for (let j = 0; j < row.length; j++) {
            const new_lines = format_argument(row[j]).split('\n');
            // new_line have at least one element that is always a string
            row[j] = new_lines.shift() as string;
            stacks.push(new_lines);
        }
        const stack_lengths = stacks.map(function(column) {
            return column.length;
        });
        const new_rows_count = Math.max.apply(Math, stack_lengths);
        for (let k = new_rows_count - 1; k >= 0; k--) {
            array.splice(i + 1, 0, stacks.map(column => column[k] ?? ''));
        }
    }
    return array as StringTable;
}

function calculate_lengths(array: StringTable) {
    return array[0].map(function(_, i) {
        var col = array.map(function(row) {
            return strlen(row[i]);
        });
        return Math.max.apply(Math, col);
    });
}

function column_padding(lengths: Array<number>, row: Array<string>) {
    return row.map(function(item, i) {
        const size = strlen(item);
        if (size < lengths[i]) {
            item += ' '.repeat(lengths[i] - size);
        }
        return item;
    });
}

function expand_tabs(str: string) {
    return str.replace(/\t/g, ' '.repeat(4));
}

function escape(str: string) {
    return str.replace(/&(?![^;]+;)/g, '&amp;');
}

export function ascii_table(array: Table, header: boolean = false) {
    if (!array.length) {
        return '';
    }
    const adjusted_array = adjust_table(array);
    const lengths = calculate_lengths(adjusted_array);
    const lines = adjusted_array.map(function(row) {
        const line = column_padding(lengths, row).join(' | ');
        return '| ' + escape(line) + ' |';
    });
    var sep = '+' + lengths.map(len => '-'.repeat(len + 2)).join('+') + '+';
    if (header) {
        return sep + '\n' + lines[0] + '\n' + sep + '\n' +
            lines.slice(1).join('\n') + '\n' + sep;
    } else {
        return sep + '\n' + lines.join('\n') + '\n' + sep;
    }
}
