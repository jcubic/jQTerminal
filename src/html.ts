import $ from 'cash-dom';
import wcwidth from 'wcwidth';

// TODO: missing items
// * warn function

import { bare_text } from './sanitize';
import { email_re, format_parts_re } from './const';
import { valid_color } from './color';
import { format_split, is_formatting, length, split_characters } from './formatters/utils';
import { is_ch_unit_supported, is_css_variables_supported } from './supports';
import { warn } from './debug';

type CharWidthOptionT = {
    charWidth: number;
};

type FormatOptionT = {
    linksNoReferrer?: boolean;
    linksNoFollow?: boolean;
    allowedAttributes?: AllowedAttrArrayT;
    charWidth?: number;
    escape?: boolean;
    anyLinks?: boolean;
};

type AllowedAttrT = string | RegExp;
type AllowedAttrArrayT = Array<AllowedAttrT>;

type StringObjectT = {
    [key: string]: string;
}

type CharSpecT = {
    len: number;
    chr: string;
}

type WideCharactersT = {
    sum: number;
    len: number;
    specs: Array<CharSpecT>;
}

function safe(string: string) {
    if (!string.match(/[<>&]/)) {
        return string;
    }
    return string.replace(/&(?![^;]+;)/g, '&amp;')
        .replace(/>/g, '&gt;').replace(/</g, '&lt;');
}

function clean_data(data: string, text?: string) {
    if (data === '') {
        return text || '';
    } else {
        return data.replace(/&#93;/g, ']')
            .replace(/>/g, '&gt;')
            .replace(/</g, '&lt;')
            .replace(/"/g, '&quot;');
    }
}

function escape_html_attr(value: string) {
    return value.replace(/"/g, '&quot;');
}

function filter_attr_names(names: string[], allowed_tags: AllowedAttrArrayT): string[] {
    if (names.length && allowed_tags.length) {
        return names.filter(function(name) {
            if (name === 'data-text') {
                return false;
            }
            let allowed = false;
            const filters = allowed_tags;
            for (let i = 0; i < filters.length; ++i) {
                if (filters[i] instanceof RegExp) {
                    if ((filters[i] as RegExp).test(name)) {
                        allowed = true;
                        break;
                    }
                } else if (filters[i] === name) {
                    allowed = true;
                    break;
                }
            }
            return allowed;
        });
    }
    return [];
}

function attrs_to_string(style: string, attrs: StringObjectT, allowed_attrs: Array<AllowedAttrT>) {
    if (attrs) {
        const keys = filter_attr_names(Object.keys(attrs), allowed_attrs);
        if (keys.length) {
            return ' ' + keys.map(function(name: string) {
                let value = escape_html_attr(attrs[name]);
                if (name === 'style') {
                    // merge style attr and colors #617
                    value = value ? style + ';' + value : style;
                }
                if (!value) {
                    return name;
                }
                return name + '="' + value + '"';
            }).join(' ');
        }
    }
    if (!style) {
        return '';
    }
    return ' style="' + style + '"';
}

function is_path(url: string) {
    return url.match(/^\.{1,2}\//) ||
        url.match(/^\//) ||
        !(url.match(/\//) || url.match(/^[^:]+:/));
}

const strlen = (function() {
    if (typeof wcwidth === 'undefined') {
        return function(string: string) {
            // fix empty prompt that use 0 width space
            string = string.replace(/\u200B/g, '');
            return length(string);
        };
    } else {
        return wcwidth;
    }
})();

function style_to_string(styles: StringObjectT) {
    return Object.keys(styles).map(function(prop) {
        return prop + ':' + styles[prop];
    }).join(';');
}

function char_width_prop(len: number, options?: CharWidthOptionT) {
    return style_to_string(char_width_object(len, options));
}

function char_width_object(len: number, options?: CharWidthOptionT) {
    const result: StringObjectT = {};
    if (len === 0) {
        result['width'] = '1px';
    } else if (is_ch_unit_supported) {
        result['width'] = len + 'ch';
    } else if (!is_css_variables_supported) {
        if (options?.charWidth) {
            result['width'] = (options.charWidth * len) + 'px';
        }
    } else {
        result['--length'] = len.toString();
    }
    return result;
}

function extra_css(text: string, options: CharWidthOptionT) {
    if (typeof wcwidth !== 'undefined') {
        const bare = bare_text(text);
        const len = strlen(bare);
        if (len > 1 && len !== length(bare)) {
            return char_width_object(len, options);
        }
    }
}

function wide_characters(text: string, options: CharWidthOptionT) {
    if (typeof wcwidth !== 'undefined') {
        const bare = bare_text(text);
        const chars = split_characters(bare);
        if (chars.length === 1) {
            return text;
        }
        const specs = chars.map(function(chr: string) {
            return {
                len: strlen(chr),
                chr: chr
            };
        }).reduce(function(arr: Array<WideCharactersT>, spec: CharSpecT) {
            const last = arr[arr.length - 1];
            if (last) {
                if (last.len !== spec.len) {
                    return arr.concat([{
                        sum: spec.len,
                        len: spec.len,
                        specs: [spec]
                    }]);
                } else {
                    arr.pop();
                    return arr.concat([{
                        sum: last.sum + spec.len,
                        len: last.len,
                        specs: last.specs.concat(spec)
                    }]);
                }
            }
            return [{
                sum: spec.len,
                specs: [spec],
                len: spec.len
            }];
        }, []);
        return specs.map(function(spec: WideCharactersT) {
            if (spec.len === 1) {
                return make_string(spec);
            }
            const style = char_width_prop(spec.sum, options);
            if (spec.sum === chars.length || !style.length) {
                return '<span>' + make_string(spec) + '</span>';
            } else if (spec.specs.length > 1) {
                return wrap(style, spec.specs.map(function(spec) {
                    return wrap(char_width_prop(spec.len), spec.chr);
                }).join(''));
            } else {
                return wrap(style, make_string(spec));
            }
        }).join('');
    }
    function make_string(spec: WideCharactersT) {
        return spec.specs.map(function(spec: CharSpecT) {
            return spec.chr;
        }).join('');
    }
    function wrap(style: string, str: string) {
        return '<span style="' + style + '">' + str + '</span>';
    }
    return text;
}

export function format(str: string, options: FormatOptionT) {
    const settings = $.extend({}, {
        linksNoReferrer: false,
        linksNoFollow: false,
        allowedAttributes: [],
        charWidth: undefined,
        escape: true,
        anyLinks: false
    }, options || {});

    function rel_attr() {
        const rel = ['noopener'];
        if (settings.linksNoReferrer) {
            rel.unshift('noreferrer');
        }
        if (settings.linksNoFollow) {
            rel.unshift('nofollow');
        }
        return rel;
    }

    // -----------------------------------------------------------------
    function with_url_validation(fn: (arg: string) => boolean) {
        return function(url: string) {
            if (settings.anyLinks) {
                return true;
            }
            const test = fn(url);
            if (!test) {
                warn('Invalid URL ' + url + ' only http(s) ftp and Path ' +
                     'are allowed');
            }
            return test;
        };
    }
    // -----------------------------------------------------------------
    const valid_href = with_url_validation(function(url) {
        return !!(url.match(/^((https?|file|ftp):\/\/|\.{0,2}\/)/) || is_path(url));
    });
    // -----------------------------------------------------------------
    const valid_src = with_url_validation(function(url) {
        return !!(url.match(/^(https?:|file:|blob:|data:)/) || is_path(url));
    });
    // -----------------------------------------------------------------
    function format(s: string, style: string, color: string, background: string, _class: string, data_text: string, text: string) {
        function pre_process_link(data: string) {
            let result;
            if (data.match(email_re)) {
                result = '<a href="mailto:' + data + '"';
            } else {
                // only http and ftp links (prevent javascript)
                // unless user force it with anyLinks option
                if (!valid_href(data)) {
                    data = '';
                }
                result = '<a target="_blank"';
                if (data) {
                    result += ' href="' + data + '"';
                }
                result += ' rel="' + rel_attr().join(' ') + '"';
            }
            return result;
        }
        function pre_process_image(data: string) {
            let result = '<img';
            if (valid_src(data)) {
                result += ' src="' + data + '"';
                if (text) {
                    result += ' alt="' + text + '"';
                }
            }
            return result;
        }
        let attrs;
        if (data_text.match(/;/)) {
            try {
                const splitted = data_text.split(';');
                const str = splitted.slice(1).join(';')
                    .replace(/&nbsp;/g, ' ')
                    .replace(/&lt;/g, '<')
                    .replace(/&gt;/g, '>');
                if (str.match(/^\s*\{[^}]*\}\s*$/)) {
                    attrs = JSON.parse(str);
                    data_text = splitted[0];
                }
            } catch (e) {
                // ignore
            }
        }
        if (text === '' && !style.match(/@/)) {
            return ''; //'<span>&nbsp;</span>';
        }
        text = safe(text);
        text = text.replace(/\\\]/g, '&#93;');
        if (settings.escape) {
            // inside formatting we need to unescape escaped slashes
            // but this escape is not needed when echo - don't know why
            text = text.replace(/\\\\/g, '\\');
        }
        const styles: StringObjectT = {};
        if (style.indexOf('b') !== -1) {
            styles['font-weight'] = 'bold';
        }
        const text_decoration = [];
        if (style.indexOf('u') !== -1) {
            text_decoration.push('underline');
        }
        if (style.indexOf('s') !== -1) {
            text_decoration.push('line-through');
        }
        if (style.indexOf('o') !== -1) {
            text_decoration.push('overline');
        }
        if (text_decoration.length) {
            styles['text-decoration'] = text_decoration.join(' ');
        }
        if (style.indexOf('i') !== -1) {
            styles['font-style'] = 'italic';
        }
        if (valid_color(color)) {
            $.extend(styles, {
                'color': color,
                '--color': color,
                '--original-color': color
            });
            if (style.indexOf('!') !== -1) {
                styles['--link-color'] = color;
            }
            if (style.indexOf('g') !== -1) {
                styles['text-shadow'] = '0 0 5px ' + color;
            }
        }
        if (valid_color(background)) {
            $.extend(styles, {
                'background-color': background,
                '--background': background
            });
        }
        const data = clean_data(data_text, text);
        const extra = extra_css(text, settings);
        if (extra) {
            text = wide_characters(text, settings);
            $.extend(styles, extra);
        }
        let result;
        if (style.indexOf('!') !== -1) {
            result = pre_process_link(data);
        } else if (style.indexOf('@') !== -1) {
            result = pre_process_image(data);
        } else {
            result = '<span';
        }
        const style_str = style_to_string(styles);
        result += attrs_to_string(style_str, attrs, settings.allowedAttributes);
        if (_class !== '') {
            result += ' class="' + _class + '"';
        }
        // links and image need data-text attribute cmd click behavior
        // formatter can return links.
        if (style.indexOf('!') !== -1) {
            result += ' data-text>' + text + '</a>';
        } else if (style.indexOf('@') !== -1) {
            result += ' data-text/>';
        } else {
            result += ' data-text="' + data + '">' +
                '<span>' + text + '</span></span>';
        }
        return result;
    }
    if (typeof str === 'string') {
        // support for formating foo[[u;;]bar]baz[[b;#fff;]quux]zzz
        const splitted = format_split(str);
        str = splitted.map(function(text: string) {
            if (text === '') {
                return text;
            } else if (is_formatting(text)) {
                // fix &nbsp; inside formatting because encode is called
                // before format
                text = text.replace(/\[\[[^\]]+\]/, function(text) {
                    return text.replace(/&nbsp;/g, ' ');
                });
                return text.replace(format_parts_re, format);
            } else {
                text = safe(text);
                text = text.replace(/\\\]/, '&#93;');
                const data = clean_data(text);
                const extra = extra_css(text, settings);
                let prefix;
                if (extra) {
                    text = wide_characters(text, settings);
                    prefix = '<span style="' + style_to_string(extra) + '"';
                } else {
                    prefix = '<span';
                }
                return prefix + ' data-text="' + data + '">' + text + '</span>';
            }
        }).join('');
        return str.replace(/<span><br\s*\/?><\/span>/gi, '<br/>');
    } else {
        return '';
    }
}
