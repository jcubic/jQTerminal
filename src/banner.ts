import version from './version';

const version_re = new RegExp(` {${version.length}}$`);

const copyrights = [
    'Copyright (c) 2023 Jakub T. Jankiewicz <https://jcu.bi>',
    '(c) 2023 Jakub T. Jankiewicz <https://jcu.bi>',
    '(c) 2023 Jakub T. Jankiewicz',
    '(c) 2023 jcubic'
].map(input => {
    const len = length(input);
    return [len, input];
});

const banners = [
    `
    _____                 ________                              __
   / _  /__  __ ___ _____/__  ___/__ ___ ______ __ __  __ ___  / /
  / // // / / // _// _  /  / // _  // _//     // //  \\/ // _ \\/ /
 / ___// /_/ // / / ___/  / // ___// / / / / // // /\\  // // / /__
/_/   /_____//_/ /____/  /_//____//_/ /_/ /_//_//_/ /_/ \\__\\_\\___/
                                                                 `,
    `
    _____                 ________
   / _  /__  __ ___ _____/__  ___/__ ___ ______
  / // // / / // _// _  /  / // _  // _//     /
 / ___// /_/ // / / ___/  / // ___// / / / / /
/_/   /_____//_/ /____/  /_//____//_/ /_/ /_/
                                            `,
    `
    _____
   / _  /__  __ ___ _____
  / // // / / // _// _  /
 / ___// /_/ // / / ___/
/_/   /_____//_/ /____/
 ________
/__  ___/__ ___ ______
  / // _  // _//     /
 / // ___// / / / / /
/_//____//_/ /_/ /_/
                   `,
    `
Pure Terminal
             `
].map(input => {
    const banner = compose(input);
    const len = length(banner);
    return [len, banner];
});

function compose(banner: string) {
    return banner.replace(/^\n/, '').replace(version_re, version);
}

function length(str: string) {
    const lines = str.split('\n');
    return Math.max(...lines.map((line: string) => line.length));
}

export const banner = (cols: number) => {
    const banner = banners.find(([len]) => {
        if (len <= cols) {
            return true;
        }
    });
    if (banner) {
        const [len, output] = banner;
        const copy = copyrights.find(([copy_len]) => {
            if (copy_len <= len) {
                return true;
            }
        });
        if (copy) {
            return [output, copy[1]].join('\n');
        }
        return output;
    }
    return '';
};
