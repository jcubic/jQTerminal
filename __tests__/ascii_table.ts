import { strlen, ascii_table } from '../src/ascii_table';

describe('strlen', () => {
    it('should get length of the normal string', () => {
        const specs = [
            'lorem',
            'ipsum',
            'dolor',
            'sit amet'
        ];
        specs.forEach(str => {
            expect(strlen(str)).toEqual(str.length);
        });
    });
    it('should return len of text with japanese characters', () => {
        const specs = [
            'ハッカー',
            '端末',
            'プログラミング'
        ];
        specs.forEach(str => {
            expect(strlen(str)).toEqual(str.length * 2);
        });
    });
    it('should skip formatting', () => {
        const specs = [
            ['[[;red;]lorem]', 5],
            ['[[;;background;]ipsum]', 5]
        ] as const;
        specs.forEach(([str, count]) => {
            expect(strlen(str)).toEqual(count);
        });
    });
});

describe('ascii_table', () => {
    it('should return empty string for empty array', () => {
        expect(ascii_table([])).toEqual('');
    });
    it('should format basic table', () => {
        const output = ascii_table([
            ['hello', 'World', 'This', 'Is Sparta'],
            ['helloworld', 'this is sparta', 'this', 'world']
        ]);
        expect(output).toMatchSnapshot();
    });
    it('should render different data types', () => {
        const output = ascii_table([
            [10000000000, 200000000, 'hello'],
            [true, false, /foo bar/],
            ['this', 'is', 'it']
        ]);
        expect(output).toMatchSnapshot();
    });
    it('should format mixed content', () => {
        const output = ascii_table([
            ['hello', 'This', 'Is', 'プログラミング'],
            ['lorem ipsum', 'dolor', 'sit\n\namet', 'hello'],
            ['hello\tworld', 'this is sparta', 'this', 'world']
        ]);
        expect(output).toMatchSnapshot();
    });
    it('should render header', () => {
        const output = ascii_table([
            ['hello', 'This', 'Is', 'プログラミング'],
            ['lorem ipsum', 'dolor', 'sit\n\namet', 'hello'],
            ['hello\tworld', 'this is sparta', 'this', 'world']
        ], true);
        expect(output).toMatchSnapshot();
    });
    it('should render empty cells', () => {
        const output = ascii_table([
            ['hello', 'This', 'Is', 'プログラミング'],
            ['lorem ipsum', '', 'sit\n\namet', 'hello'],
            ['hello\tworld', 'this is sparta', '', '']
        ], true);
        expect(output).toMatchSnapshot();
    });
});
