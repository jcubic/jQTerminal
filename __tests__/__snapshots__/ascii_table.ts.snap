// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ascii_table should format basic table 1`] = `
"+------------+----------------+------+-----------+
| hello      | World          | This | Is Sparta |
| helloworld | this is sparta | this | world     |
+------------+----------------+------+-----------+"
`;

exports[`ascii_table should format mixed content 1`] = `
"+----------------+----------------+------+----------------+
| hello          | This           | Is   | プログラミング |
| lorem ipsum    | dolor          | sit  | hello          |
|                |                |      |                |
|                |                | amet |                |
| hello    world | this is sparta | this | world          |
+----------------+----------------+------+----------------+"
`;

exports[`ascii_table should render different data types 1`] = `
"+-------------+-----------+-----------+
| 10000000000 | 200000000 | hello     |
| true        | false     | /foo bar/ |
| this        | is        | it        |
+-------------+-----------+-----------+"
`;

exports[`ascii_table should render empty cells 1`] = `
"+----------------+----------------+------+----------------+
| hello          | This           | Is   | プログラミング |
+----------------+----------------+------+----------------+
| lorem ipsum    |                | sit  | hello          |
|                |                |      |                |
|                |                | amet |                |
| hello    world | this is sparta |      |                |
+----------------+----------------+------+----------------+"
`;

exports[`ascii_table should render header 1`] = `
"+----------------+----------------+------+----------------+
| hello          | This           | Is   | プログラミング |
+----------------+----------------+------+----------------+
| lorem ipsum    | dolor          | sit  | hello          |
|                |                |      |                |
|                |                | amet |                |
| hello    world | this is sparta | this | world          |
+----------------+----------------+------+----------------+"
`;
